### ISSUE TITLE
Custom Plugin: Accessories Resetting

### ISSUE THREAD

**Issue by Vortec4800** [homebridge/homebridge#2697]
**Describe Your Problem:**
I've written a custom plugin and it seems to work well. However, once every few days the custom accessory exposed by my plugin resets the room/favorite/status switches back to the defaults and it gets removed from any automations it was in. It's like HomeKit thinks its a new accessory.

I haven't figured out a pattern as to when this happens, and I haven't seen anything in the logs when it does either. This is really a plugin development issue but I don't know where the best place is to get support for that. Can you point me in the right direction?

**Environment:**

* **Node.js Version**: v12.18.2
* **NPM Version**: v6.14.7
* **Homebridge Version**: 1.2.3
* **Operating System**: Official Pi Image

**Comment by NorthernMan54** [comment]
Pls share a link to your github repo with the source code.

Typically this is caused by accessories being deleted and added again in the plugin.

**Comment by Vortec4800** [comment]
Sure, here it is:

https://github.com/Vortec4800/homebridge-ecobee-away

**Comment by NorthernMan54** [comment]
That is about 2 months old....

**Comment by Vortec4800** [comment]
I don't understand. I wrote it 2 months ago?

**Comment by NorthernMan54** [comment]
Is that your latest code?

**Comment by Vortec4800** [comment]
Yes, aside from this reset bug everything works fine and I haven't had to make any changes.

**Comment by NorthernMan54** [comment]
My initial thought was device discovery removing and adding the accessory again or the service constructor creating new devices occasionally, but your code looks good.

Is your .persist or .accessories directories or files being removed or ???

**Comment by Vortec4800** [comment]
I also thought maybe it was a bug in my caching code, I used it basically unchanged from the plugin template and I thought maybe I had to add something else to handle an edge case or something, but I wasn't sure.

Not that I'm aware of, I'm using the Homebridge official Pi image and I don't have anything else installed on that Pi, it's dedicated to Homebridge only. I really only manage it from the Config-X interface and it doesn't lose power or anything unusual.

I'll see what I can do about monitoring the accessory more closely for the next few days and see if I can't find out when it happens, and try to grab logs from it. I'll report back if I can do that, otherwise let me know if you think of anything else to check or try.

Worth noting, this does also happen on another camera plugin I have installed, where the cameras reset their settings all back to defaults and automations are removed as well. I have several other plugins installs and none of them have this issue, only my plugin and the camera plugin. I just barely noticed that one though, so I'm not sure yet if both plugins do it at the same time or if they're separate, but I'm monitoring that as well now too.

**Comment by ebaauw** [comment]
Maybe a corrupt file system on the micro-SD card?

**Comment by adriancable** [comment]
@Vortec4800 - there are really only two base causes of this. One is if ~/.homebridge/persist/*.json gets lost on restart. This will happen e.g. if you are using Docker, if you are restarting your Pi in an unsafe way which may not sync the filesystem, etc.

Second is if you are recreating accessories and not preserving the UUID. HomeKit will then not know it's the same accessory, and so you will lose automations, move back to the default room etc.

**Comment by Vortec4800** [comment]
Okay, this happened again last night and I just noticed it this morning. I grabbed the logs, maybe there will be something useful here.

I saw there was an update available for Config-UI-X so I did that from the web interface. Here is the log from that, including the start-up logs afterwards. The entire system didn't reboot during this, the web interface just restarted the homebridge instance like it normally does. That said I don't think it's a filesystem thing, unless the homebridge process getting terminated by config-ui-x isn't writing the json files properly or something.

```
[10/11/2020, 20:26:53] [Config] Running Command: sudo -E -n npm install homebridge-config-ui-x@latest
[10/11/2020, 20:30:34] [Config] [homebridge-config-ui-x] Failed to check registry.npmjs.org for updates: "timeout of 5000ms exceeded" - see https://git.io/JJSz6 for help.
[10/11/2020, 20:30:34] [Config] [homebridge-ecobee-away] Failed to check registry.npmjs.org for updates: "timeout of 5000ms exceeded" - see https://git.io/JJSz6 for help.
[10/11/2020, 20:30:36] [Config] Homebridge restart request received
[10/11/2020, 20:30:43] [Config] No restart command defined, killing process...
[10/11/2020, 20:31:15] [HB Supervisor] OS: Linux 5.4.51-v7l+ arm
[10/11/2020, 20:31:15] [HB Supervisor] Node.js v12.18.2 /usr/local/bin/node
[10/11/2020, 20:31:15] [HB Supervisor] Homebridge Path: /usr/local/lib/node_modules/homebridge/bin/homebridge
[10/11/2020, 20:31:15] [HB Supervisor] UI Path: /usr/local/lib/node_modules/homebridge-config-ui-x/dist/bin/standalone.js
[10/11/2020, 20:31:15] [HB Supervisor] Starting Homebridge with extra flags: -I
[10/11/2020, 20:31:15] [HB Supervisor] Started Homebridge v1.2.3 with PID: 7902
[10/11/2020, 20:31:16] Loaded config.json with 6 accessories and 7 platforms.
[10/11/2020, 20:31:16] ---
[10/11/2020, 20:31:18] Loaded plugin: homebridge-apple-tv-remote@0.12.0
[10/11/2020, 20:31:18] Registering platform 'homebridge-apple-tv-remote.AppleTvPlatform'
[10/11/2020, 20:31:18] ---
[10/11/2020, 20:31:18] Loaded plugin: homebridge-config-ui-x@4.28.1
[10/11/2020, 20:31:18] Registering platform 'homebridge-config-ui-x.config'
[10/11/2020, 20:31:18] ---
[10/11/2020, 20:31:18] Loaded plugin: homebridge-delay-switch@2.2.4
[10/11/2020, 20:31:18] Registering accessory 'homebridge-delay-switch.DelaySwitch'
[10/11/2020, 20:31:18] ---
[10/11/2020, 20:31:18] Loaded plugin: homebridge-dummy@0.4.1
[10/11/2020, 20:31:18] Registering accessory 'homebridge-dummy.DummySwitch'
[10/11/2020, 20:31:18] ---
[10/11/2020, 20:31:18] Loaded plugin: homebridge-ecobee-away@1.0.1
[10/11/2020, 20:31:18] Registering platform 'homebridge-ecobee-away.EcobeeAway'
[10/11/2020, 20:31:18] ---
[10/11/2020, 20:31:18] Loaded plugin: homebridge-elkm1@1.1.10
[10/11/2020, 20:31:18] Registering platform 'homebridge-elkm1.ElkM1'
[10/11/2020, 20:31:18] ---
[10/11/2020, 20:31:18] Loaded plugin: homebridge-lutron-caseta-smockle@3.0.59
[10/11/2020, 20:31:18] Registering platform 'homebridge-lutron-caseta-smockle.LutronCasetaPlatform'
[10/11/2020, 20:31:18] ---
[10/11/2020, 20:31:20] [Config] Homebridge Config UI X v4.28.1 is listening on :: port 8581
[10/11/2020, 20:31:19] Loaded plugin: homebridge-myq@2.3.2
[10/11/2020, 20:31:19] Registering platform 'homebridge-myq.myQ'
[10/11/2020, 20:31:19] ---
[10/11/2020, 20:31:19] Loaded plugin: homebridge-schedule@1.0.58
[10/11/2020, 20:31:19] Registering accessory 'homebridge-schedule.Schedule'
[10/11/2020, 20:31:19] ---
[10/11/2020, 20:31:19] Loaded plugin: homebridge-unifi-protect@3.7.9
[10/11/2020, 20:31:19] Registering platform 'homebridge-unifi-protect.UniFi Protect'
[10/11/2020, 20:31:19] ---
[10/11/2020, 20:31:19] Loading 7 platforms...
[10/11/2020, 20:31:19] [Config] Initializing config platform...
[10/11/2020, 20:31:19] [Config] Running in Service Mode
[10/11/2020, 20:31:19] [ELK M1] Initializing ElkM1 platform...
[10/11/2020, 20:31:19] [ELK M1] Connecting to M1
[10/11/2020, 20:31:19] [LutronCasetaPlatform] Initializing LutronCasetaPlatform platform...
[10/11/2020, 20:31:19] [Apple TV Platform] Initializing AppleTvPlatform platform...
[10/11/2020, 20:31:19] [Ecobee Away] Initializing EcobeeAway platform...
[10/11/2020, 20:31:19] [UniFi Protect] Initializing UniFi Protect platform...
[10/11/2020, 20:31:19] [myQ] Initializing myQ platform...
[10/11/2020, 20:31:19] Loading 6 accessories...
[10/11/2020, 20:31:19] [Lock All Doors Trigger] Initializing DelaySwitch accessory...
[10/11/2020, 20:31:19] [ELK Disarmed] Initializing DummySwitch accessory...
[10/11/2020, 20:31:19] [ELK Disarmed] Setting switch to true
[10/11/2020, 20:31:19] [ELK Away] Initializing DummySwitch accessory...
[10/11/2020, 20:31:19] [ELK Away] Setting switch to false
[10/11/2020, 20:31:19] [ELK Home] Initializing DummySwitch accessory...
[10/11/2020, 20:31:19] [ELK Home] Setting switch to false
[10/11/2020, 20:31:19] [Night] Initializing DummySwitch accessory...
[10/11/2020, 20:31:19] [Night] Setting switch to true
[10/11/2020, 20:31:19] [Hourly] Initializing Schedule accessory...
[10/11/2020, 20:31:19] [Hourly] Creating schedule accessory [Hourly]
[10/11/2020, 20:31:19] [Hourly] Initialization complete
[10/11/2020, 20:31:19] [Hourly] Sarting [0 * * * *] cron job
[10/11/2020, 20:31:20] [Apple TV Platform] Initialing platform...
[10/11/2020, 20:31:20] [Apple TV Platform] [Master Bathroom Apple TV] Initializing...
[10/11/2020, 20:31:20] [Ecobee Away] Renewing auth token
[10/11/2020, 20:31:20] [ELK M1] ***Connected***
Enter this code with your HomeKit app on your iOS device to pair with Homebridge:
                       
    ┌────────────┐     
    └────────────┘     
                       
[10/11/2020, 20:31:21] Homebridge v1.2.3 is running on port 51377.
[10/11/2020, 20:31:21] [myQ] myQ API: Successfully connected to the myQ API.
[10/11/2020, 20:31:21] [Ecobee Away] Adding new accessory: Ecobee Away
```

Of note is the very last log, which is the one from my custom plugin's startup sequence. The normal restart log from the plugin is:

```
[Ecobee Away] Restoring existing accessory from cache: Ecobee Away
```

However it looks like Homebridge doesn't know that the accessory already exists, so it's creating it from scratch.

**Comment by Vortec4800** [comment]
Just discovered something else weird that might be a clue, I recently added the MyQ plugin as well (within the past week or so) and my garage door was also reset/rediscovered. It's like the device cache is reverting back to an older state without my most recently installed plugins.

**Comment by Supereg** [comment]
`configureAccessory` is never actually called, meaning there is no entry in the accessory cache. You said it happens only sometimes? the cache should file is located in the homebridge directory under the `accessories` folder.

**Comment by Vortec4800** [comment]
That's correct, this doesn't happen every time and I still haven't been able to track down a pattern as to when it does happen.

I'm not aware of any plugins that are doing anything weird with the cache, meaning I don't think I'm using any obscure plugins that don't have issues like this normally. My custom plugin is also pretty straight-forward.

Here is the listing for the accessories folder, which also looks normal.

```bash
pi@homebridge:/var/lib/homebridge/accessories $ ls -l
total 116
-rw-r--r-- 1 pi pi 108724 Oct 12 16:43 cachedAccessories
-rw-r--r-- 1 pi pi   6103 Sep  6 13:20 uiAccessoriesLayout.json
```

**Comment by jgraessley** [comment]
I had this same problem in a plug-in I wrote and just realized my mistake. The sample my plug-in is based on had this line for updating the platform accessories:
this.api.updatePlatformAccessories([existingAccessory]);

I switched it to call this.api.updatePlatformAccessories(this.accessories);

In my case my platform is finding (and restoring) 4 accessories every time, but for each accessory it was calling this.api.updatePlatformAccessories([existingAccessory]). It seemed like that caused the problem, maybe because I was inadvertently telling homekit my plugins entire set of accessories was that single accessory? I'm new to Javascript/Typescript so I'm not sure how to find documentation on updatePlatformAccessories to know if I'm expected to always past the full list of accessories my plugin is responsible for every time or just those that may have changed. Hope this helps.
